enhancements/cluster-logging/cluster-logging-es-pipeline-processing.md

```diff
@@ -0,0 +1,205 @@
+
+However, common data model does not solve another challenge associated with the ask to properly index, store and query any additional (extra) custom data the documents can carry. For example logs can contain project or business domain specific JSON data that is not captured in the common data model and is not shared by all the data sources.
+
+Over the time we have gathered number of requests from customers asking for possibility to process additional data. See <https://issues.redhat.com/browse/LOG-785>.
+
+For the purpose of this proposal we have already investigated various approaches (and their pros and cons) of storing additional JSON data in Elasticsearch indices as part of <https://issues.redhat.com/browse/LOG-842>.
+
+
+### Goals
+
+The high-level goal of this proposal is to lay the groundwork to enable approaches B) and C) as discussed in [LOG-842](https://issues.redhat.com/browse/LOG-842).
+
+Quick comparison of approaches B) and C):
+
+- Approach B) Whenever user needs to store data with a new data model we solve it by pointing this data to new index (or to existing index if there is already one with exactly the same data model). This is very comfy approach for the user except it can lead to very high number of indices and querying experience can be complicated because there are very little or no common conventions of document fields meaning and types.
+
+- Approach C) is on the other side of the spectrum. All data is stored into common index and whenever user needs to store data with a new data model then it is necessary to transform the data to fit existing data model in existing index. This approach requires attention to data model and users may need to invest time and resources to transform new data sources. On the other hand there is minimum additional indices created in ES cluster and data can be queried uniformly because document fields have predictable meaning and types.
+
+It should be pointed out that Elasticsearch pipeline processes can support both cases. Depending on how much users are willing to pay attention to their data model they can start with approach B) and ventually transition to approach C) once they feel more confident about managing their own data model across more log resources.
+
+In this proposal we assume that the approach B is the one we would start enabling first. But nothing is stopping us from enabling the approach C in the future. Though it is fair to say that approach C will be more complicated to implement because it will require management of custom processing pipelines.
```

Comment:
given the potential for having lots of indices and difficulty querying for customers, why would we want to implement approach B?

Reply:
Approach C) requires better understanding of what needs to be done to avoid index explosion.

The whole concept is as follows:
1) if you have an extra JSON data then you need to route the logs to a new index (very simple, short time win)
2) if you run into high number of indices you need to start thinking how to share common index between individual projects
  2.1) this can mean you need to change how you create logs (not likely)
  2.2) or you need to change logs before they are pushed into ES, and this is where we are starting getting into approach C), because now it means you need to think about data model and how you unify them
3) If you want to query (and analyze) logs across indices with different models it will be challenging
  3.1) you will need to start thinking about how to unify data models to be able to formulate meaningful ES queries and this will lead you to unifying the models again...

It will start with approach B) but will need to continue towards approach C).
Should we target C) from the beginning? We could but this still means we will need the B) as well. So for those users that can be happy with just B) they can get it earlier.

Reply:
This proposal does not go into details how approach C) should be implemented. This should be subject of follow up proposal but this will have to reflect our "lessons learned" from more specific use cases. Approach C) will be built on top of B) anyway.

Reply:
Can we update the language in the proposal here to denote that C will be built on top of B.

Can we also change this from "B" and "C" to "A" and "B" since there are only two proposals being outlined here?

Reply:
@ewolinetz The "A", "B", "C" and "D" refers to equally named approaches investigated in https://issues.redhat.com/browse/LOG-842
Hence, I would like to keep it in sync.

Reply:
> Can we update the language in the proposal here to denote that C will be built on top of B.

I am now saying this:
"... This also means that implementing the approach C) requires implement all that is needed for the approach B) first anyway."